vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc

```diff
@@ -0,0 +1,194 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/confirmations/internal/payments.h"
+#include "bat/confirmations/internal/static_values.h"
+#include "bat/confirmations/internal/logging.h"
+#include "bat/confirmations/internal/confirmations_impl.h"
+
+#include "base/json/json_reader.h"
+
+namespace confirmations {
+
+Payments::Payments(
+    ConfirmationsImpl* confirmations,
+    ConfirmationsClient* confirmations_client) :
+    payments_({}),
+    confirmations_(confirmations),
+    confirmations_client_(confirmations_client) {
+  BLOG(INFO) << "Initializing payments";
+  (void)confirmations_;
+}
+
+Payments::~Payments() {
+  BLOG(INFO) << "Deinitializing payments";
+}
+
+bool Payments::ParseJson(const std::string& json) {
+  base::Optional<base::Value> value = base::JSONReader::Read(json);
+  if (!value || !value->is_list()) {
+    return false;
+  }
+
+  base::ListValue* list = nullptr;
+  if (!value->GetAsList(&list)) {
+    return false;
+  }
+
+  payments_ = GetPaymentsFromList(list);
+
+  return true;
+}
+
+double Payments::GetBalance() const {
+  double balance = 0.0;
+
+  for (const auto& payment : payments_) {
+    balance += payment.balance;
+  }
+
+  return balance;
+}
+
+base::Time Payments::CalculateNextPaymentDate(
+    uint64_t next_token_redemption_date_in_seconds) const {
+  auto now = base::Time::Now();
+  base::Time::Exploded now_exploded;
+  now.LocalExplode(&now_exploded);
+
+  int month = now_exploded.month;
+
+  if (now_exploded.day_of_month <= kNextPaymentDay) {
+    if (HasPendingBalanceForLastMonth()) {
+      // If last month has a pending balance, then the next payment date will
+      // occur this month
+    } else {
+      // If last month does not have a pending balance, then the next payment
+      // date will occur next month
+      month++;
+    }
+  } else {
+    if (HasPendingBalanceForThisMonth()) {
+      // If this month has a pending balance, then the next payment date will
+      // occur next month
+      month++;
+    } else {
+      auto next_token_redemption_date =
+          base::Time::FromDoubleT(next_token_redemption_date_in_seconds);
+
+      base::Time::Exploded next_token_redemption_date_exploded;
+      next_token_redemption_date.LocalExplode(
+          &next_token_redemption_date_exploded);
+
+      if (next_token_redemption_date_exploded.month == month) {
+        // If this month does not have a pending balance and our next token
+        // redemption date is this month, then the next payment date will occur
+        // next month
+        month++;
+      } else {
+        // If this month does not have a pending balance and our next token
+        // redemption date is next month, then the next payment date will occur
+        // the month after next
+        month += 2;
+      }
+    }
+  }
+
+  int year = now_exploded.year;
+
+  if (month > 12) {
+    month -= 12;
+    year++;
+  }
+
+  std::string date = std::to_string(year);
```

Comment:
Using `base::StringPrintf` here would probably be a little simpler.

Reply:
thanks just what I was originally looking for :-)

Reply:
Fixed